reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt

```diff
@@ -0,0 +1,214 @@
+    }
+
+    @Test
+    fun completes_WHEN_upstream_completed_without_values() {
+        val observer = testConcatMap { TestObservable() }
+
+        upstream.onComplete()
+
+        assertTrue(observer.isCompleted)
+    }
+
+    @Test
+    fun does_not_complete_WHEN_upstream_produced_values_and_completed() {
+        val observer = testConcatMap { TestObservable() }
+
+        upstream.onNext(0)
+        upstream.onComplete()
+
+        assertFalse(observer.isCompleted)
+    }
+
+    @Test
+    fun completes_WHEN_upstream_produced_values_and_completed_and_all_sources_are_completed() {
+        val inners = innerSources(3)
+        val observer = testConcatMap(inners)
+
+        upstream.onNext(0)
+        inners[0].onComplete()
+        upstream.onNext(1)
+        inners[1].onComplete()
+        upstream.onNext(2)
+        upstream.onComplete()
+        inners[2].onComplete()
+
+        assertTrue(observer.isCompleted)
+    }
+
+    @Test
+    fun does_not_complete_WHEN_upstream_produced_values_and_completed_and_not_all_inner_sources_are_completed() {
+        val inners = innerSources(3)
+        val observer = testConcatMap(inners)
+
+        upstream.onNext(0)
+        inners[0].onComplete()
+        upstream.onNext(1)
+        upstream.onNext(2)
+        upstream.onComplete()
+        inners[1].onComplete()
+
+        assertFalse(observer.isCompleted)
+    }
+
+    @Test
+    fun produces_error_WHEN_inner_source_produced_error() {
+        val inner = TestObservable<String>()
+        val observer = testConcatMap { inner }
+
+        upstream.onNext(0)
+        inner.onError(Throwable())
+
+        assertTrue(observer.isError)
+    }
+
+    @Test
+    fun does_not_produce_more_values_WHEN_disposed() {
+        val inner = TestObservable<String>()
+        val observer = testConcatMap { inner }
+        upstream.onNext(0)
+        inner.onNext("a")
+        observer.reset()
+
+        observer.dispose()
+        inner.onNext("b")
+
+        assertFalse(observer.hasOnNext)
+    }
+
+    @Test
+    fun disposes_source_WHEN_disposed() {
+        val inner = TestObservable<String>()
+        val observer = testConcatMap { inner }
+        upstream.onNext(0)
+
+        observer.dispose()
+
+        assertTrue(inner.isDisposed)
+    }
+
+    @Test
+    fun disposes_source_WHEN_upstream_produced_error() {
+        val inner = TestObservable<String>()
+        testConcatMap { inner }
+        upstream.onNext(0)
+
+        upstream.onError(Throwable())
+
+        assertTrue(inner.isDisposed)
+    }
+
+    @Test
+    fun disposes_upstream_WHEN_source_produced_error() {
+        val inner = TestObservable<String>()
+        testConcatMap { inner }
+        upstream.onNext(0)
+
+        inner.onError(Throwable())
+
+        assertTrue(upstream.isDisposed)
+    }
+
+    private fun testConcatMap(innerSources: List<Observable<String?>>): TestObservableObserver<String?> =
```

Comment:
It's not clear what these methods do. Perhaps subscribeWithIndexedObservableList?